drake/core/Vector.h

```diff
@@ -0,0 +1,269 @@
+      };
+    };
+
+    template<template<typename> class Vector1,
+            template<typename> class Vector2>
+    struct CombinedVectorHelper<Vector1, Vector2, true> {
+      // Vec1 non-empty, Vec2 empty
+      template<typename ScalarType> using type = Vector1<ScalarType>;
+
+      template<typename ScalarType>
+      static const Vector1<ScalarType> &first(const Vector1<ScalarType> &vec) { return vec; }
+
+      template<typename ScalarType>
+      static NullVector <ScalarType> second(const Vector1<ScalarType> &vec) { return NullVector<ScalarType>(); }
+
+      template<typename ScalarType>
+      static Vector1<ScalarType> combine(const Vector1<ScalarType> &vec1,
+                                         const Vector2<ScalarType> &vec2) { return vec1; };
+    };
+  }
+
+  /** CombinedVectorUtil
+   * @brief provides logic to build combined vectors and access the first and second elements of the combined vector, handling the \
+  case when the combined vector builder could have returned the original type
+   *
+   * Uses template aliasing so that combining a vector with the NullVector simply returns the orginal vector type.
+   */
+  template<template<typename> class Vector1,
+          template<typename> class Vector2, typename Vec1IsNull = void>
+  struct CombinedVectorUtil {
+    // Vec1 is non-empty (Vec2 could be either)
+    typedef typename internal::template CombinedVectorHelper<Vector1, Vector2,
+            Vector2<double>::RowsAtCompileTime == 0> helper;
+    template<typename ScalarType> using type = typename helper::template type<ScalarType>;
+
+    template<typename ScalarType>
+    static Vector1<ScalarType> first(const typename helper::template type<ScalarType> &vec) { return helper::first(vec); }
+
+    template<typename ScalarType>
+    static Vector2<ScalarType> second(const typename helper::template type<ScalarType> &vec) {
+      return helper::second(vec);
+    }
+
+    template<typename ScalarType>
+    static typename helper::template type<ScalarType> combine(const Vector1<ScalarType> &vec1,
+                                                              const Vector2<ScalarType> &vec2) {
+      return helper::combine(vec1, vec2);
+    }
+  };
+
+  template<template<typename> class Vector1,
+          template<typename> class Vector2>
+  struct CombinedVectorUtil<Vector1, Vector2, typename std::enable_if<Vector1<double>::RowsAtCompileTime == 0>::type> {
+    // Vec1 is empty (Vec2 could be either)
+    template<typename ScalarType> using type = Vector2<ScalarType>;
+
+    template<typename ScalarType>
+    static NullVector <ScalarType> first(const Vector2<ScalarType> &vec) { return NullVector<ScalarType>(); }
+
+    template<typename ScalarType>
+    static const Vector2<ScalarType> &second(const Vector2<ScalarType> &vec) { return vec; }
+
+    template<typename ScalarType>
+    static Vector2<ScalarType> combine(const Vector1<ScalarType> &vec1, const Vector2<ScalarType> &vec2) { return vec2; };
+  };
```

Comment:
We can talk about this later, but I think the CombinedVectorUtil / CombinedVectorHelper stuff is too complex; I don't understand it and was running into issues with it when trying to do a dynamic sized system. I'll open an issue on that when I get back to the Python-wrappable dynamic size system.